hack/Makefile.debug

```diff
@@ -2,29 +2,137 @@
+# allows for remote debugging. Attach from your IDE at any time after
+# launch to begin debugging.
+DLV_CONTINUE ?= --continue
+
+build:
 	GO111MODULE=on CGO_ENABLED=0 GOFLAGS=-mod=vendor go build -o openshift-router -gcflags=all="-N -l" ./cmd/openshift-router
-	$(IMAGEBUILDER) build -t $(IMAGE):$(TAG) -f hack/Dockerfile.debug .
+	GO111MODULE=on CGO_ENABLED=0 GOFLAGS=-mod=vendor go build -o hack/remote-debug/remote-debug-helper -gcflags=all="-N -l" ./hack/remote-debug/remote-debug-helper.go
+
+build-image: build
+	$(IMAGEBUILDER) build --build-arg HAPROXY_RPMS="$(HAPROXY_RPMS)" --build-arg GITHUB_USER="$(GITHUB_USER)" -t $(IMAGE):$(TAG) -f $(DOCKERFILE_DEBUG) .
 
-push:
+push-image:
 	$(IMAGEBUILDER) tag $(IMAGE):$(TAG) $(REGISTRY)/$(IMAGE):$(TAG)
 	$(IMAGEBUILDER) push $(REGISTRY)/$(IMAGE):$(TAG)
 
 set-image:
 	oc scale --replicas 0 -n openshift-cluster-version deployments/cluster-version-operator
 	oc scale --replicas 0 -n openshift-ingress-operator deployments ingress-operator
-	oc -n openshift-ingress scale deployment --replicas=0 router-default
-	oc -n openshift-ingress patch deployment router-default -p '{"spec":{"template":{"spec":{"$$setElementOrder/containers":[{"name":"router"}],"containers":[{"imagePullPolicy":"Always","name":"router"}]}}}}'
+	oc scale --replicas 0 -n openshift-ingress deployment/router-default
+ifeq ($(DOCKERFILE_DEBUG),hack/remote-debug/Dockerfile)
+	oc apply -f hack/remote-debug/manifests
+	oc apply -f hack/remote-debug/custom-router-scc.yaml
+	@for patch in $(sort $(wildcard hack/remote-debug/patches/deployment/router-default/*.yaml)); do \
+	    echo "Applying patch $$patch"; \
+	    oc -n openshift-ingress patch deployment router-default -p "$$(cat $$patch)" --type=strategic; \
+	done
+endif
 	oc -n openshift-ingress set image deployment/router-default router=$(REGISTRY)/$(IMAGE):$(TAG)
-	oc -n openshift-ingress scale deployment --replicas=1 router-default
+	oc scale --replicas 1 -n openshift-ingress deployment/router-default
+ifeq ($(DOCKERFILE_DEBUG),hack/remote-debug/Dockerfile)
+	@echo "Waiting for the deployment/router-default to become available..."; \
+	if ! oc -n openshift-ingress rollout status deployment/router-default --timeout=180s; then \
+		echo "Deployment did not become available in time. Please check the deployment status."; \
+		false; \
+	fi
+	@max_attempts=10; \
+	attempt=1; \
+	while [ $$attempt -le $$max_attempts ]; do \
+		echo "Checking SSH access..."; \
+		if ssh -o ConnectTimeout=5 -F hack/remote-debug/ssh_config container uptime; then \
+			echo "SSH check passed."; \
+			break; \
+		else \
+			echo "SSH check failed. Restarting port-forward services and retrying..."; \
+			$(MAKE) -f hack/Makefile.debug restart-port-forward-services; \
+			attempt=$$((attempt + 1)); \
+			sleep 5; \
+		fi; \
+	done; \
+	if [ $$attempt -gt $$max_attempts ]; then \
+		echo "SSH check did not pass in time. Please check the deployment and SSH configuration."; \
+		false; \
+	fi
```

Comment:
It ensures the port forward service is working.

Reply:
> It ensures the port forward service is working.

So does the readiness probe, no? I think something like this can be a replacement for both `oc rollout status` and this check:
```
oc -n openshift-ingress wait --for=condition=Available=true deploy/router-default --timeout=180s
```

Reply:
I don't think readiness proves that ssh works. The readiness probe will succeed (it's hitting the ssh port, and will immediately close) but (as I just found out) that's no guarantee that you can genuinely login.